Make CI fail if Clippy finds any warnings.

crsaracco opened the pull request:
Fail the CI tests if Clippy reports any warnings.

askeksa commented on src/util/parameter_transfer.rs:
```diff
@@ -180,7 +182,7 @@ mod tests {
         // Verify final values
         for p in 0..PARAMETERS {
-            assert!((0..THREADS).any(|t| results[t][p] == values[p]));
+            assert!((0..THREADS).any(|t| (results[t][p] - values[p]).abs() < std::f32::EPSILON));
```
The test must use actual equality here, not the approximate one. It tests that the exact value present was set by at least one thread.

crsaracco replied:
Done (along with other fuzzy float compares)

askeksa commented on src/util/test_util.rs:
```diff
@@ -0,0 +1,13 @@
+/// Compare two `f32`s, and assert equality.
+pub fn assert_f32_equal(a: f32, b: f32) {
+    assert!((a - b).abs() < std::f32::EPSILON);
+}
```
Many float comparisons have been changed to compare explicitly with EPSILON. If we have a function for it, we should use that in all relevant places. Also, rename the function to make it clear that the comparison is approximate.

askeksa commented on src/util/test_util.rs:
```diff
@@ -0,0 +1,13 @@
+/// Assert that an `f32` is zero.
+pub fn assert_f32_zero(a: f32) {
+    assert!(a.abs() < std::f32::EPSILON);
+}
```
This function seems redundant (and has a misleading name), since comparing against 0.0 using the approximate comparison function would be at least as clear.

askeksa commented on src/buffer.rs:
```diff
@@ -495,86 +504,77 @@ mod tests {
-                assert_eq!(*output, 0.0);
+            input.iter().zip(output.iter_mut()).fold(0, |acc, (input, output)| {
+                test_util::assert_f32_equal(*input, acc as f32);
+                test_util::assert_f32_zero(*output);
```
The output was set to 0.0 explicitly, and the assert checks that it was not changed. It would seem that an exact comparison is more appropriate here.

askeksa commented on src/buffer.rs:
```diff
@@ -593,15 +593,12 @@ mod tests {
-                assert_eq!(*output, 0.0);
+            input.iter().zip(output.iter_mut()).fold(0, |acc, (input, output)| {
+                test_util::assert_f32_equal(*input, acc as f32);
+                test_util::assert_f32_zero(*output);
```
The output was set to 0.0 explicitly, and the assert checks that it was not changed. It would seem that an exact comparison is more appropriate here.

crsaracco commented:
@askeksa changes SGTM, I'll re-visit them and fix after #102 is merged in.

askeksa commented:
Actually, when looking closer, it looks to me like none of the approximate float comparisons are warranted. They are either checking if some value has changed since last time, checking against a specific value set in a test, or comparing with a value obtained by summing some integers. In all of these cases, exact comparison is the right thing to do.

crsaracco commented:
> Actually, when looking closer, it looks to me like none of the approximate float comparisons are warranted. They are either checking if some value has changed since last time, checking against a specific value set in a test, or comparing with a value obtained by summing some integers. In all of these cases, exact comparison is the right thing to do.

Agreed. I think I was just blindly following clippy for that lint when I made this PR.

Reverted back to strict compares, added `#[allow(clippy::float_cmp)]` for those tests, and removed the `test_util` module I added since the functions weren't being used anymore.

crsaracco commented on examples/transfer_and_smooth.rs:
```diff
@@ -75,7 +75,7 @@ impl Plugin for MyPlugin {
             name: "transfer_and_smooth".to_string(),
             vendor: "Loonies".to_string(),
-            unique_id: 0x500007,
+            unique_id: 0x0050_0007,
```
Yeah, I kinda agree (could take them or leave them). I reverted most of these, and added a crate-wide allow in our CI file.

crsaracco commented on examples/ladder_filter.rs:
```diff
@@ -190,7 +190,7 @@ impl PluginParameters for LadderParameters {
             2 => format!("{}", self.poles.load(Ordering::Relaxed) + 1),
             3 => format!("{:.3}", self.drive.get()),
-            _ => format!(""),
+            _ => "".to_string(),
```
For some reason, it seems like clippy isn't complaining about this anymore, so reverted.

crsaracco commented on src/buffer.rs:
```diff
@@ -493,6 +493,7 @@ mod tests {
     /// and the output channels are just 0.
     /// This test assures that when the buffers are zipped together,
     /// the input values do not change.
+    #[allow(clippy::float_cmp)]
```
We can add it crate-wide through the CI file.

For right now, I'm hesitant to ignore this particular one crate-wide in case we actually *do* do some strict float compares where they shouldn't be, and having the `allow` statements makes it easy to audit in the future.

Although on the other hand, I can't really think of a scenario where a strict compare would be incorrect and a fuzzy float compare would be correct, when talking about the internals of a VST crate. So... idk?

askeksa replied:
I think the set of enforced lints should only contain ones that are usually a good idea, and where there could be a few exceptions. `float_cmp` is rather the other way around. There are only false positives, and quite a lot of them.

askeksa replied:
Also, I don't find it likely that we would ever need fuzzy float comparisons in this crate, outside of examples. Whereas the exact comparisons are quite proliferate in tests.

crsaracco replied:
Yeah, I can get behind this as of right now. Might change my mind if we ever do come across a `float_cmp` that we should have caught :wink:

crsaracco commented:
> Many of the Clippy suggestions are good, but some are really bad, making me wonder if enforcing no Clippy warnings across the repo is such a good idea after all. At least without a generous, crate-wide list of exceptions (which, it seems, is not yet possible).

I don't know, I think the Clippy suggestions are good enough to warrant enforcement in the general case, especially in the context of an open source project where anyone can make a PR. Even moreso when [other projects in the ecosystem use Clippy to help identify ecosystem-wide issues](https://github.com/rust-secure-code/safety-dance/issues/21).

I think it's easy enough to chuck on a `#[allow(clippy::whatever)]` with justifications in the code review (I would even like the justifications in the comments, but anyways...)

Finally, if a particular lint is causing us too much trouble for what it's worth, we can decide to ignore it crate-wide in the CI file, like I did for "unreadable_literal"s in this PR.

askeksa commented:
And yes, I do agree that it makes sense to turn the warnings on, with appropriate exceptions. I think we should have the three mentioned. Two of them because they are "causing us too much trouble for what it's worth" as you say, and one because it is buggy (which together with just one false positive warrants disabling, IMO).

crsaracco commented on src/util/parameter_transfer.rs:
```diff
@@ -181,8 +181,9 @@ mod tests {
         assert!(transfer.iterate(true).next().is_none());
 
         // Verify final values
+        #[allow(clippy::needless_range_loop)]
```
Agreed. I made this a crate-wide allow for now. I've subscribed to that bug and will make another PR to remove that allow when that bug is resolved.

crsaracco commented:
So just to be explicit, the three crate-wide ignored warnings are going to be (as of this PR):

 - `clippy::unreadable_literal`
 - `clippy::needless_range_loop`
 - `clippy::float_cmp`

Right?

askeksa commented:
Yes, exactly.

crsaracco commented on src/plugin.rs:
```diff
@@ -406,6 +406,8 @@ pub enum CanDo {
 
 impl CanDo {
+    // TODO: either rename this function or implement FromStr
+    #![allow(clippy::should_implement_trait)]
```
Would require a slight signature change, so I'm deferring to #110. Should be pretty easy though.
